File: src/c02_oop/matlab/Association.m
function Association()
    % Explanation
    % Composition:
    % Car contains an Engine object. The Engine is created when the Car is constructed,
    % and it can't exist outside of it.
    %
    % Aggregation:
    % Car has a Driver object. This allows the Driver to exist independently and
    % be shared with other instances of Car.

    % Composition: Car contains an Engine
    myCar = Car('Toyota', 'V8');

    % Aggregation: Driver is shared and can be assigned to different cars
    driver = Driver('Alice');
    myCar.setDriver(driver);

    myCar.start(); % Start the car with the assigned driver
end

classdef Engine < handle
    properties (Access = private)
        model
    end
    methods
        function obj = Engine(model)
            obj.model = model;
        end
        function start(obj)
            fprintf('Engine %s started.\n', obj.model);
        end
        function m = getModel(obj)
            m = obj.model;
        end
    end
end

classdef Driver < handle
    properties (Access = private)
        name
    end
    methods
        function obj = Driver(name)
            obj.name = name;
        end
        function drive(obj)
            fprintf('Driver %s is driving.\n', obj.name);
        end
        function n = getName(obj)
            n = obj.name;
        end
    end
end

classdef Car < handle
    properties (Access = private)
        brand
        engine % Composition: Car *owns* the engine
        driver % Aggregation: Car *uses* the driver
    end
    methods
        function obj = Car(brand, engineModel)
            obj.brand = brand;
            obj.engine = Engine(engineModel);
        end
        function setDriver(obj, driver)
            obj.driver = driver;
        end
        function start(obj)
            fprintf('Car %s is starting.\n', obj.brand);
            obj.engine.start();
            if ~isempty(obj.driver)
                obj.driver.drive();
            else
                fprintf('No driver assigned.\n');
            end
        end
    end
end

File: src/c02_oop/Association.java
package c02_oop;

/**
 * Explanation
 * Composition:
 *
 * Car contains an Engine object. The Engine is created when the Car is
 * constructed,
 * and it can't exist outside of it.
 *
 * Aggregation:
 *
 * Car has a Driver object. This allows the Driver to exist independently and
 * be shared with other instances of Car.
 */
public class Association {

    static class Engine {
        private String model;

        public Engine(String model) {
            this.model = model;
        }

        public void start() {
            System.out.println("Engine " + model + " started.");
        }

        public String getModel() {
            return model;
        }
    }

    static class Driver {
        private String name;

        public Driver(String name) {
            this.name = name;
        }

        public void drive() {
            System.out.println("Driver " + name + " is driving.");
        }

        public String getName() {
            return name;
        }
    }

    static class Car {
        private String brand;
        private Engine engine; // Composition: Car *owns* the engine
        private Driver driver; // Aggregation: Car *uses* the driver

        public Car(String brand, String engineModel) {
            this.brand = brand;
            this.engine = new Engine(engineModel);
        }

        public void setDriver(Driver driver) {
            this.driver = driver;
        }

        public void start() {
            System.out.println("Car " + brand + " is starting.");
            engine.start();
            if (driver != null) {
                driver.drive();
            } else {
                System.out.println("No driver assigned.");
            }
        }
    }

    public static void main(String[] args) {
        // Composition: Car contains an Engine
        Car myCar = new Car("Toyota", "V8");

        // Aggregation: Driver is shared and can be assigned to different cars
        Driver driver = new Driver("Alice");
        myCar.setDriver(driver);

        myCar.start(); // Start the car with the assigned driver
    }
}
